src/preprocessing/passes/normalize.cpp

```diff
@@ -0,0 +1,1063 @@
+          normalized[bound_vars] = normalizedBoundVarList;
+        }
+
+        // Push unvisited children onto the stack
+        for (int i = current.getNumChildren() - 1; i >= 0; --i)
+        {
+          Node child = current[i];
+          if (visited.find(child) == visited.end())
+          {
+            stack.push(child);
+          }
+        }
+
+        // For APPLY_UF nodes, push the operator
+        if (current.getKind() == cvc5::internal::Kind::APPLY_UF)
+        {
+          Node op = current.getOperator();
+          if (visited.find(op) == visited.end())
+          {
+            stack.push(op);
+          }
+        }
+
+        // Leave 'it->second' as false; we'll process this node later
+      }
+    }
+    else if (!it->second)
+    {
+      // Second time seeing this node, process it after its children
+
+      std::vector<Node> children;
+
+      if (current.getKind() == cvc5::internal::Kind::APPLY_UF)
+      {
+        // Handle operator for APPLY_UF nodes
+        auto opIt = normalized.find(current.getOperator());
+        Assert(opIt != normalized.end());
+        children.push_back(opIt->second);
+      }
+      else if (current.getMetaKind() == metakind::PARAMETERIZED)
+      {
+        // For parameterized nodes, include the operator
+        children.push_back(current.getOperator());
+      }
+
+      for (size_t i = 0; i < current.getNumChildren(); ++i)
+      {
+        Node child = current[i];
+        auto childIt = normalized.find(child);
+        Assert(childIt != normalized.end());
+        children.push_back(childIt->second);
+      }
+
+      Node ret = nodeManager->mkNode(current.getKind(), children);
+      normalized[current] = ret;
+
+      // Mark as processed and pop from the stack
+      it->second = true;
+      stack.pop();
+    }
+    else
+    {
+      // Node has already been processed
+      stack.pop();
+    }
+  }
+
+  return normalized[n];
+}
+
+/**
+ * Checks if a node represents a trivial equality (e.g., x = x).
+ *
+ * @param n The node to check.
+ * @return True if the node is a trivial equality, false otherwise.
+ */
+bool isTrivialEquality(const Node& n)
+{
+  if (n.getKind() == cvc5::internal::Kind::EQUAL)
+  {
+    const auto &lhs = n[0], rhs = n[1];
+    if (lhs == rhs)
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Checks if a node represents the constant true.
+ *
+ * @param n The node to check.
+ * @return True if the node is the constant true, false otherwise.
+ */
+bool isTrue(const Node& n)
+{
+  if (n.isConst() && n.getType().isBoolean())
+  {
+    return n.getConst<bool>();
+  }
+  return false;
+}
+
+/**
+ * Collects all unique types present in a node's subtree.
+ *
+ * @param n The root node of the subtree.
+ * @param types Vector to store the collected types.
+ * @param visited Set to track visited nodes.
+ * @param mark Set to track collected types.
+ */
+void collectTypes(TNode n,
```

Comment:
Why not just use the standard method https://github.com/cvc5/cvc5/blob/main/src/expr/node_algorithm.h#L256?